internal/pkg/stream/cfnstackset.go

```diff
@@ -0,0 +1,150 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package stream
+
+import (
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws/request"
+
+	"github.com/aws/copilot-cli/internal/pkg/aws/cloudformation/stackset"
+)
+
+// StackSetDescriber is the CloudFormation interface needed to describe the health of a stack set operation.
+type StackSetDescriber interface {
+	InstanceSummaries(name string, opts ...stackset.InstanceSummariesOption) ([]stackset.InstanceSummary, error)
+	DescribeOperation(name, opID string) (stackset.Operation, error)
+}
+
+// StackSetOpEvent represents a stack set operation status update message.
+type StackSetOpEvent struct {
+	Name      string // The name of the stack set.
+	Operation stackset.Operation
+}
+
+// StackSetStreamer is a [Streamer] emitting [StackSetOpEvent] messages for instances under modification.
+type StackSetStreamer struct {
+	stackset    StackSetDescriber
+	ssName      string
+	opID        string
+	opStartTime time.Time
+
+	mu         sync.Mutex
```

Comment:
is this only for subs? if so, can we call it 
```suggestion
	subsMu         sync.Mutex
```
or just add a doc string saying it protects `subs`?
?

Reply:
Done!